aten/src/ATen/native/TensorIteratorReduce.cpp

```diff
@@ -28,21 +30,33 @@ static bool use_two_pass_reduction(TensorIterator& iter) {
 }
 
 static void two_pass_reduction(TensorIterator& iter, const loop2d_t& loop) {
-  int num_threads = at::get_max_threads();
+  int max_threads = at::get_max_threads();
 
   auto& dst = iter.tensor(0);
   auto buffer_shape = DimVector(dst.sizes());
-  buffer_shape.insert(buffer_shape.begin(), num_threads);
+  buffer_shape.insert(buffer_shape.begin(), max_threads);
   auto buffer = at::empty(buffer_shape, dst.type());
 
+  std::unique_ptr<bool[]> written(new bool[max_threads]);
+  std::fill(written.get(), written.get() + max_threads, false);
+
   at::parallel_for(0, iter.numel(), internal::GRAIN_SIZE, [&](int64_t begin, int64_t end) {
-    auto slice = buffer[at::get_thread_num()];
+    int thread_num = at::get_thread_num();
+    written[thread_num] = true;
```

Comment:
That doesn’t look thread safe. The compiler could happily emit stores to 4B values here and make threads clobber each other’s outputs. Can you make those atomic?

Reply:
No, that's not correct. Each byte in a byte array has a different memory location. Concurrent writes to different memory locations are allowed with no synchronization requirements.